### DISCUSSION TITLE
"imputed population" and "Sample QC flags"

### DISCUSSION THREAD

**Discussion by jxchong** [broadinstitute/seqr#2417] [Q&A]
How do we add/edit metadata/values to the "Imputed Population" and "Sample QC flags" sections for each sample? These would be quite useful for us. I searched for those phrases in the repository but only came up with constants.js which wasn't informative. Also, for Sample QC flags, can we add arbitrary key/values or are there pre-specific values? Thanks!

**Comment by hanars** [comment]
This data is added through the "Data Management" -> "Sample QC" interface, and it expects the output of @mike-w-wilson sample QC pipeline that he separately runs on our data. I'm not sure if the code to generate the sample qc file is ready to be open sourced yet, but we did write the upload parsing and display logic to be very tightly coupled with this pipeline so if its not able to be shared yet we would probably need to update seqr if you wanted to use these fields. If it is ready to be shared then I will let you know

**Reply by jxchong** [reply]
Thanks. Would it be possible to shoehorn other data into the input pipeline? We generate these values already through our own pipeline and could write some scripts to transform it to match.

**Reply by hanars** [reply]
yes, I think shoehorning your data in should be fine. You can provide a tsv/csv/excel file with the following columns, let me know if you have any questions:

- **data_type**
WES or WGS. Must be the same for all rows in the file. If you want to load qc data for both exome and genome data do it as 2 separate file uploads
- **seqr_id**
The ID for the corresponding seqr individual. That individual must already have data loaded in seqr of the specified data_type 
- **filter_flags**
JSON representation of a list of filter flags (i.e. `["chimera", "coverage"]`). Flags can be any of the following values, and must have values in the corresponding columns (i.e. if `chimera` is in the list it will pull the value from the `AL_PCT_CHIMERAS` column)
    ```
    'callrate': 'filtered_callrate',
    'contamination': 'PCT_CONTAMINATION',
    'chimera': 'AL_PCT_CHIMERAS',
    'coverage': if exome data - 'HS_PCT_TARGET_BASES_20X', if genome data 'WGS_MEAN_COVERAGE'
    ``` 
- **qc_metrics_filters**
JSON representation of a list of qc metric flags (i.e. `["r_insertion_deletion", "n_not_called"]`). Flags can represent any ratio or numeric value that is an outlier in sample qc. Fields beginning with `r_` are displayed as ratios, fields beginning with `n_` are displayed as numbers (i.e. `r_insertion_deletion` will be shown as "Ratio Insertion/Deletion" while `n_not_called` will be shown as "Num. Not Called"). All flags must have a corresponding column prefaced with `sample_qc.` (so if `r_insertion_deletion` is included there needs to be a `sample_qc.r_insertion_deletion` column)
- **qc_pop**
Imputed population. The following are the populations we use which will display prettily in seqr, but if you provide a different value it will be displayed as whatever you set (so if you use 'AFR' it will show up in seqr as "African" but if you use `Pacific Islander` it will just show up in seqr as "Pacific Islander")
  ```
  AFR: African
  AMR: Latino
  ASJ: Ashkenazi Jewish
  EAS: East Asian
  FIN: European (Finnish)
  MDE: Middle Eastern
  NFE: European (non-Finnish)
  OTH: Other
  SAS: South Asian
  ```

Optional columns. These are required if any of the filters/ flags corresponding to them are included
- filtered_callrate - number
- PCT_CONTAMINATION - number
- AL_PCT_CHIMERAS - number
- HS_PCT_TARGET_BASES_20X - number
- WGS_MEAN_COVERAGE - number
- sample_qc.* - columns correspond to  all possible values for `qc_metrics_filters`. The value is not shown in seqr so it doesn;t really matter, but these are usually numeric

Example display:
![image](https://user-images.githubusercontent.com/24598672/151410449-c4fd86e6-ebfb-461e-806f-dfbb2c35e685.png)

**Reply by jxchong** [reply]
Thanks! If we upload a second file with a subset of flags/values, does it entirely replace all values for that sample or only the values that are present in the new file?

**Reply by hanars** [reply]
entirely replaces
